src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2022 Justin Thomas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Justin Thomas
+-/
+import data.polynomial
+import data.polynomial.ring_division
+import ring_theory.principal_ideal_domain
+import algebra.module.linear_map
+import field_theory.minpoly
+import linear_algebra
+import ring_theory.ideal.operations
+import ring_theory.polynomial_algebra
+
+/-!
+# Annihilating Ideal
+
+Given a commutative ring `R` and an `R`-algebra `A`
+Every element `a : A` defines
+an ideal `polynomial.annihilating_ideal a ⊆ R[X]`.
+Simply put, this is the set of polynomials `p` where
+the polynomial evaluation `p(a)` is 0.
+
+## Special case where the ground ring is a field
+
+In the special case that `R` is a field, we use the notation `R = 𝕜`.
+Here `𝕜[X]` is a PID, so there is a polynomial `g ∈ polynomial.annihilating_ideal a`
+which generates the ideal. We show that if this generator is
+chosen to be monic, then it is the minimal polynomial of `a`,
+as defined in `field_theory.minpoly`.
+
+## Special case: endomorphism algebra
+
+Given an `R`-module `M` (`[add_comm_group M] [module R M]`)
+there are some common specialiazations which may be more familiar.
+* Example 1: `A = M →ₗ[R] M`, the endomorphism algebra of an `R`-module M.
+* Example 2: `A = n × n` matrices with entries in `R`.
+-/
+
+open_locale polynomial
+
+namespace polynomial
+
+section general_case
```

Comment:
I think in `mathlib` it's common just to write the requirements in the `section` names, so for example this would be `semiring`, `field_case` would be just `field`, and so on

Reply:
will be fixed in next push. removed minpoly_generates section since it was based on the goal lemma instead of requirements